content/en/blog/_posts/2020-09-02-scaling-kubernetes-networking-endpointslices.md

```diff
@@ -0,0 +1,46 @@
+
+By default, EndpointSlices  store as many as 100 endpoints each, though this can be configured with the `--max-endpoints-per-slice` flag on kube-controller-manager.
+
+## EndpointSlices provide 10x scalability improvements
+This API dramatically improves networking scalability. Now when a Pod is added or removed, only 1 small EndpointSlice needs to be updated. This difference becomes quite noticeable when hundreds or thousands of Pods are backing a single Service. 
+
+Potentially more significant, now that all Pod IPs for a Service don't need to be stored in a single resource, we don't have to worry about the size limit for objects stored in etcd. EndpointSlices have already been used to scale Services beyond 100,000 network endpoints.
+
+All of this is brought together with some significant performance improvements that have been made in kube-proxy. When using EndpointSlices at scale, significantly less data will be transferred for endpoints updates and kube-proxy should be faster to update iptables or ipvs rules. Beyond that, Services can now scale to at least 10 times beyond any previous limitations.
+
+## EndpointSlices enable new functionality
+Introduced as an alpha feature in Kubernetes v1.16, EndpointSlices were built to enable some exciting new functionality in future Kubernetes releases. This could include dual-stack Services, topology aware routing, and endpoint subsetting.
+
+Dual-Stack Services are an exciting new feature that has been in development alongside EndpointSlices. They will utilize both IPv4 and IPv6 addresses for Services and rely on the addressType field on EndpointSlices to track these addresses by IP family. 
+
+Topology aware routing will update kube-proxy to prefer routing requests within the same zone or region. This makes use of the topology fields stored for each endpoint in an EndpointSlice. As a further refinement of that, we're exploring the potential of endpoint subsetting. This would allow kube-proxy to only watch a subset of EndpointSlices. For example, this might be combined with topology aware routing so that kube-proxy would only need to watch EndpointSlices containing endpoints within the same zone. This would provide another very significant scalability improvement.
+
+## What does this mean for the Endpoints API?
+Although the EndpointSlice API is providing a newer and more scalable alternative to the Endpoints API, the Endpoints API will continue to be considered generally available and stable. The most significant change planned for the Endpoints API will involve beginning to truncate Endpoints that would otherwise run into scalability issues.
+
+The Endpoints API is not going away, but many new features will rely on the EndpointSlice API. To take advantage of the new scalability and functionality that EndpointSlices provide, applications that currently consume Endpoints will likely want to consider switching to EndpointSlices in the future.
```

Comment:
we might want to encourage them to switch _once endpointslices reach v1_... the effort to switch twice (to beta, then to v1) might be more than we want to push

Reply:
That's a good point. What are the pain points involved in transitioning from v1beta1 to v1? Would it just be updating some client/informer calls?

I'm a bit hesitant to suggest that all implementations wait until EndpointSlices reach v1. I think it could be helpful to have some more early adopters of the API so that we can smooth out any pain points in time for the v1 API. My goal here was to be relatively vague - encouraging adoption of the API in the future without being too specific or pushing for a certain transition timeline.

When an application starts to support EndpointSlices will likely be dependent on the user base and implementation goals. In many cases, applications will go through a phase where they support both APIs (assuming they support a wide range of Kubernetes versions). For those cases, getting initial opt-in implementations in place sooner than later could be helpful. That's especially true if it helps us uncover any shortcomings of the API before it graduates to v1.

Do you think there's a better way to communicate this? Not quite sure what updated wording might look like here without getting too detailed or prescriptive.

Reply:
If they actually keep supporting/fallback to endpoints/v1 in parallel, that seems ok, but I've seen components switch to a new beta API without really considering scenarios where the beta API is not available.

Reply:
That's a good point, maybe a better way to say this is "consider supporting EndpointSlices in the future" instead of "switching to".

Reply:
edit this last sentence @robscott @liggitt